### DISCUSSION TITLE
Escape, or Authorization message handling

### DISCUSSION THREAD

**Discussion by tloose** [FortySevenEffects/arduino_midi_library#338] [Q&A]
Has anyone had any experience de4aling with "Authorization" or "Escape" MIDI messages?  I believe they **start** with 0xF7.

**Comment by franky47** [comment]
0xF7 is a End of System Exclusive marker, so I doubt a valid message starts with that.

What device is this intended for?

**Reply by tloose** [reply]
From the Standard MIDI File Format Spec doc:

Another form of sysex event is provided which does not imply that an F0 should be transmitted. This may be used as an "escape" to providef or the transmission of things which would not otherwise be legal, including system real time messages, song pointer or select, MIDI Time Code, etc. This uses the F7 code:
F7 <length> <all bytes to be transmitted>

**Reply by franky47** [reply]
I see, this library does not conform to the MIDI **File** format, but to the MIDI stream format for communications with MIDI hardware using DIN connectors (or more modern transports like USB or BLE).

As such, it does not support this format of message.

**Reply by tloose** [reply]
I am attempting to record MIDI sequences from a Rodgers organ across it's sequencer MIDI ports.  Sometimes it works, sometimes it does not.  When it does not work the MIDI file has weird escape sequences in it where I was expecting a sysex message.  I wrote a simulator for the organ to allow me to send arbitrary MIDI sequences (using the sendSysEx() function in the MIDI library) using a second cpu board and MIDI hardware.  When I send a sequence starting with 0xF7 I get the same type of weird escape sequence.
I think the organ is sending 0xF7 messages sometimes, and 0xF0 Sysex messages sometimes.

**Comment by tloose** [comment]
Are you saying this type of message should never be seen in a MIDI stream?

**Reply by franky47** [reply]
Yes, it's a MIDI file specific message. The [MIDI file specification](http://www.music.mcgill.ca/~ich/classes/mumt306/StandardMIDIfileformat.html) says (emphasis mine):

> The specification defines the 8-bit binary data stream **used in the file**. The data can be stored in a binary file, nibbilized, 7-bit-ized for efficient MIDI transmission, converted to Hex ASCII, or translated symbolically to a printable text file. This spec addresses what's in the 8-bit stream. **It does not address how a MIDI File will be transmitted over MIDI**.

Transmitting a MIDI file as-is on the wire will result in invalid MIDI data (escape as you describe, but also meta-events which would result in a SystemReset event).

**Comment by tloose** [comment]
I get the distinction between how the MIDI file is encoded and what appears
across the MIDI cable.  I've been trying to figure out what the
Rodgers/Roland instrument is doing that causes a problem during recording
by observing the effects; unsuccessfully so far.  I am going to the
instrument today  with some tools to look at the raw MIDI stream from the
instrument under specific operating conditions that are causing problems.
I hope to know more soon.

On Wed, Oct 18, 2023 at 7:10 AM François Best ***@***.***>
wrote:

> Yes, it's a MIDI file specific message. The MIDI file specification
> <http://www.music.mcgill.ca/~ich/classes/mumt306/StandardMIDIfileformat.html>
> says (emphasis mine):
>
> The specification defines the 8-bit binary data stream *used in the file*.
> The data can be stored in a binary file, nibbilized, 7-bit-ized for
> efficient MIDI transmission, converted to Hex ASCII, or translated
> symbolically to a printable text file. This spec addresses what's in the
> 8-bit stream. *It does not address how a MIDI File will be transmitted
> over MIDI*.
>
> Transmitting a MIDI file as-is on the wire will result in invalid MIDI
> data (escape as you describe, but also meta-events which would result in a
> SystemReset event).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FortySevenEffects/arduino_midi_library/discussions/338#discussioncomment-7314523>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AT34RNOKM4CXW22JIZRF3PLX77BKZAVCNFSM6AAAAAA6EMH6YCVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TGMJUGUZDG>
> .
> You are receiving this because you authored the thread.Message ID:
> <FortySevenEffects/arduino_midi_library/repo-discussions/338/comments/7314523
> @github.com>
>


-- 
TimL

**Comment by tloose** [comment]
Thanks for responding to this discussion.

I was able to run some more experiments with the Rodgers/Roland organ,
including looking at the raw data from the MIDI out port.

It is *always* sending properly formatted MIDI messages, including Sysex
messages with the correct start and stop bytes.

I do still see some malformed messages when I use MIDI recording functions;
No proper MIDI headers, etc.  I am now suspicious of something like an
uninitialized buffer, pointer out of bounds, etc.

If I may ask for some advice:  what functions from the library should I
invoke before recording to a file, and when I start recording, and to
finish and close recording, and to prepare to record again?

On Wed, Oct 18, 2023 at 7:10 AM François Best ***@***.***>
wrote:

> Yes, it's a MIDI file specific message. The MIDI file specification
> <http://www.music.mcgill.ca/~ich/classes/mumt306/StandardMIDIfileformat.html>
> says (emphasis mine):
>
> The specification defines the 8-bit binary data stream *used in the file*.
> The data can be stored in a binary file, nibbilized, 7-bit-ized for
> efficient MIDI transmission, converted to Hex ASCII, or translated
> symbolically to a printable text file. This spec addresses what's in the
> 8-bit stream. *It does not address how a MIDI File will be transmitted
> over MIDI*.
>
> Transmitting a MIDI file as-is on the wire will result in invalid MIDI
> data (escape as you describe, but also meta-events which would result in a
> SystemReset event).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FortySevenEffects/arduino_midi_library/discussions/338#discussioncomment-7314523>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AT34RNOKM4CXW22JIZRF3PLX77BKZAVCNFSM6AAAAAA6EMH6YCVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TGMJUGUZDG>
> .
> You are receiving this because you authored the thread.Message ID:
> <FortySevenEffects/arduino_midi_library/repo-discussions/338/comments/7314523
> @github.com>
>


-- 
TimL
